<?xml version="1.0" ?>
<robot name="rover">
  <!-- Constants -->
  <!-- Inertial Macros-->
  <!-- Joints/Links Macros -->
  <!-- Joints/Links -->
  <link name="base_link"/>
  <link name="chassis_link">
    <inertial>
      <mass value="15"/>
      <inertia ixx="0.53798625" ixy="0" ixz="0" iyy="0.32746625" iyz="0" izz="0.78173"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="../meshes/rover_chassis.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="-0.085906 0 0"/>
      <geometry>
        <box size="0.277 0.63 0.183"/>
      </geometry>
    </collision>
    <collision>
      <origin xyz="0.163609 0 -0.049449"/>
      <geometry>
        <box size="0.167 0.63 0.0891 "/>
      </geometry>
    </collision>
  </link>
  <link name="arm_base_link">
    <inertial>
      <mass value="3.0"/>
      <inertia ixx="102.5" ixy="0.0" ixz="0.0" iyy="24.5" iyz="0.0" izz="102.5"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../meshes/rover_chassis.dae"/>
      </geometry>
      <material name="red">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
  </link>
  <link name="arm_a_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="16.4882 20.0235 49.1172"/> -->
      <geometry>
        <mesh filename="../meshes/arm_a_fix.dae"/>
      </geometry>
      <material name="white">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
  </link>
  <link name="arm_b_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="19.923 20.023 54.02"/> -->
      <geometry>
        <mesh filename="../meshes/arm_b_fix.dae"/>
      </geometry>
      <material name="orange">
        <color rgba="1 0.43 0 1"/>
      </material>
    </visual>
  </link>
  <link name="arm_c_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="73.3593 20.0235 54.9252"/> -->
      <geometry>
        <mesh filename="../meshes/arm_c_fix.dae"/>
      </geometry>
      <material name="yellow">
        <color rgba="1 1 0 1"/>
      </material>
    </visual>
  </link>
  <link name="arm_d_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="127.963 20.0235 63.7846"/> -->
      <geometry>
        <mesh filename="../meshes/arm_d_fix.dae"/>
      </geometry>
      <material name="green">
        <color rgba="0.43 1 0 1"/>
      </material>
    </visual>
  </link>
  <link name="arm_e_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="132.451 20.0235 63.7846"/> -->
      <geometry>
        <mesh filename="../meshes/arm_e_fix.dae"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
  </link>
  <link name="arm_gripper_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <!-- <origin rpy="0 0 0" xyz="132.451 20.0235 63.7846"/> -->
      <geometry>
        <mesh filename="../meshes/arm_gripper_fix.dae"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
  </link>
  <joint name="chassis_to_arm_a" type="prismatic">
    <origin rpy="0 0 0" xyz="16.4882 20.0235 49.1172"/>
    <parent link="chassis_link"/>
    <child link="arm_a_link"/>
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-40" upper="0" velocity="0"/>
  </joint>
  <joint name="arm_a_to_arm_b" type="revolute">
    <origin rpy="0 0 0" xyz="3.4348 -0.0005 4.9028"/>
    <parent link="arm_a_link"/>
    <child link="arm_b_link"/>
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.7853981875" upper="0" velocity="0"/>
  </joint>
  <joint name="arm_b_to_arm_c" type="revolute">
    <origin rpy="0 0 0" xyz="53.4363 0.0005 0.9052"/>
    <parent link="arm_b_link"/>
    <child link="arm_c_link"/>
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.959931" upper="2.87979" velocity="0"/>
  </joint>
  <joint name="arm_c_to_arm_d" type="revolute">
    <origin rpy="0 0 0" xyz="54.6037 0 8.8594"/>
    <parent link="arm_c_link"/>
    <child link="arm_d_link"/>
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-2.35619" upper="2.35619" velocity="0"/>
  </joint>
  <joint name="arm_d_to_arm_e" type="revolute">
    <origin rpy="0 0 0" xyz="4.488 0 0"/>
    <parent link="arm_d_link"/>
    <child link="arm_e_link"/>
    <axis xyz="1 0 0"/>
    <limit effort="0" lower="-2.35619" upper="2.35619" velocity="0"/>
  </joint>
  <joint name="gripper_link" type="prismatic">
    <origin rpy="1.570796 0 0" xyz="0.0875 0 0"/>
    <parent link="arm_e_link"/>
    <child link="arm_gripper_link"/>
    <axis xyz="0 1 0"/>
    <limit lower="0" upper="8" effort="0" velocity="0"/>
    <dynamics damping="1.0"/>
  </joint>
  
  <!-- BELOW IS NEW SHIT I HAVENT FIXED -->

  <link name="front_left_wheel_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.014275000000000001" ixy="0" ixz="0" iyy="0.014275000000000001" iyz="0" izz="0.02535"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796 0 0" xyz="0 29 53.5"/>
      <geometry>
        <mesh filename="../meshes/rover_left_wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.08" radius="0.13"/>
      </geometry>
    </collision>
  </link>
  <link name="center_left_wheel_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.014275000000000001" ixy="0" ixz="0" iyy="0.014275000000000001" iyz="0" izz="0.02535"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796 0 0" xyz="-39 29 53.5"/>
      <geometry>
        <mesh filename="../meshes/rover_left_wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.08" radius="0.13"/>
      </geometry>
    </collision>
  </link>
  <link name="back_left_wheel_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.014275000000000001" ixy="0" ixz="0" iyy="0.014275000000000001" iyz="0" izz="0.02535"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796 0 0" xyz="-83 29 53.5"/>
      <geometry>
        <mesh filename="../meshes/rover_left_wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.08" radius="0.13"/>
      </geometry>
    </collision>
  </link>
  <link name="left_bogie_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.125" ixy="0" ixz="0" iyy="0.125" iyz="0" izz="0.125"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="../meshes/rover_left_bogie.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 -0.734" xyz="-0.14 0.1 0"/>
      <geometry>
        <box size="0.2 0.025 0.025"/>
      </geometry>
    </collision>
    <collision>
      <origin rpy="0 0 0.734" xyz="0.1 0.088 0"/>
      <geometry>
        <box size="0.2 0.025 0.025"/>
      </geometry>
    </collision>
  </link>
  <link name="left_rocker_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.125" ixy="0" ixz="0" iyy="0.125" iyz="0" izz="0.125"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="../meshes/rover_left_rocker.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0.585" xyz="0.11 0.077 0"/>
      <geometry>
        <box size="0.2 0.025 0.025"/>
      </geometry>
    </collision>
    <collision>
      <origin rpy="0 0 -0.585" xyz="-0.22 0.15 0"/>
      <geometry>
        <box size="0.5 0.025 0.025"/>
      </geometry>
    </collision>
  </link>
  <!-- <joint name="chassis_link_to_left_rocker_link" type="revolute">
    <parent link="chassis_link"/>
    <child link="left_rocker_link"/>
    <origin rpy="-1.570796375 0 0" xyz="0.164665 0.365 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.122173" upper="0.523599" velocity="0"/>
  </joint>
  <joint name="left_rocker_link_to_back_left_wheel_link" type="continuous">
    <parent link="left_rocker_link"/>
    <child link="back_left_wheel_link"/>
    <origin xyz="-0.417094 0.28513 0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="2.68" velocity="0"/>
  </joint>
  <joint name="left_rocker_link_to_left_bogie_link" type="revolute">
    <parent link="left_rocker_link"/>
    <child link="left_bogie_link"/>
    <origin xyz="0.198948 0.138674 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.349066" upper="0.349066" velocity="0"/>
  </joint>
  <joint name="left_bogie_link_to_front_left_wheel_link" type="continuous">
    <parent link="left_bogie_link"/>
    <child link="front_left_wheel_link"/>
    <origin xyz="0.215233 0.147933 0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="2.68" velocity="0"/>
  </joint>
  <joint name="left_bogie_link_to_center_left_wheel_link" type="continuous">
    <parent link="left_bogie_link"/>
    <child link="center_left_wheel_link"/>
    <origin xyz="-0.173532 0.147659 0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="2.68" velocity="0"/>
  </joint> -->
  <link name="front_right_wheel_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.014275000000000001" ixy="0" ixz="0" iyy="0.014275000000000001" iyz="0" izz="0.02535"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="../meshes/rover_right_wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.08" radius="0.13"/>
      </geometry>
    </collision>
  </link>
  <link name="center_right_wheel_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.014275000000000001" ixy="0" ixz="0" iyy="0.014275000000000001" iyz="0" izz="0.02535"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="-39 0 0"/>
      <geometry>
        <mesh filename="../meshes/rover_right_wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.08" radius="0.13"/>
      </geometry>
    </collision>
  </link>
  <link name="back_right_wheel_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.014275000000000001" ixy="0" ixz="0" iyy="0.014275000000000001" iyz="0" izz="0.02535"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="-83 0 0"/>
      <geometry>
        <mesh filename="../meshes/rover_right_wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.08" radius="0.13"/>
      </geometry>
    </collision>
  </link>
  <link name="right_bogie_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.125" ixy="0" ixz="0" iyy="0.125" iyz="0" izz="0.125"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="../meshes/rover_right_bogie.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 -0.734" xyz="-0.14 0.1 0"/>
      <geometry>
        <box size="0.2 0.025 0.025"/>
      </geometry>
    </collision>
    <collision>
      <origin rpy="0 0 0.734" xyz="0.1 0.088 0"/>
      <geometry>
        <box size="0.2 0.025 0.025"/>
      </geometry>
    </collision>
  </link>
  <link name="right_rocker_link">
    <inertial>
      <mass value="3"/>
      <inertia ixx="0.125" ixy="0" ixz="0" iyy="0.125" iyz="0" izz="0.125"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="../meshes/rover_right_rocker.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0.585" xyz="0.11 0.077 0"/>
      <geometry>
        <box size="0.2 0.025 0.025"/>
      </geometry>
    </collision>
    <collision>
      <origin rpy="0 0 -0.585" xyz="-0.22 0.15 0"/>
      <geometry>
        <box size="0.5 0.025 0.025"/>
      </geometry>
    </collision>
  </link>
  <!-- <joint name="chassis_link_to_right_rocker_link" type="revolute">
    <parent link="chassis_link"/>
    <child link="right_rocker_link"/>
    <origin rpy="-1.570796375 0 0" xyz="0.164665 -0.365 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.122173" upper="0.523599" velocity="0"/>
  </joint>
  <joint name="right_rocker_link_to_back_right_wheel_link" type="continuous">
    <parent link="right_rocker_link"/>
    <child link="back_right_wheel_link"/>
    <origin xyz="-0.417094 0.28513 -0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="2.68" velocity="0"/>
  </joint>
  <joint name="right_rocker_link_to_right_bogie_link" type="revolute">
    <parent link="right_rocker_link"/>
    <child link="right_bogie_link"/>
    <origin xyz="0.198948 0.138674 0"/>
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.349066" upper="0.349066" velocity="0"/>
  </joint>
  <joint name="right_bogie_link_to_front_right_wheel_link" type="continuous">
    <parent link="right_bogie_link"/>
    <child link="front_right_wheel_link"/>
    <origin xyz="0.215233 0.147933 -0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="2.68" velocity="0"/>
  </joint>
  <joint name="right_bogie_link_to_center_right_wheel_link" type="continuous">
    <parent link="right_bogie_link"/>
    <child link="center_right_wheel_link"/>
    <origin xyz="-0.173532 0.147659 -0.09"/>
    <axis xyz="0 0 1"/>
    <limit effort="2.68" velocity="0"/>
  </joint>
  <joint name="chassis_link_to_left_gps" type="fixed">
    <origin xyz="0 0.1 0"/>
    <parent link="chassis_link"/>
    <child link="left_gps"/>
  </joint> -->
  <link name="left_gps"/>
  <joint name="chassis_link_to_right_gps" type="fixed">
    <origin xyz="0 -0.1 0"/>
    <parent link="chassis_link"/>
    <child link="right_gps"/>
  </joint>
  <link name="right_gps"/>
  <joint name="chassis_link_to_center_gps" type="fixed">
    <origin xyz="0 0 1"/>
    <parent link="chassis_link"/>
    <child link="center_gps"/>
  </joint>
  <link name="center_gps"/>
  <joint name="chassis_link_to_imu" type="fixed">
    <origin xyz="0 0 0"/>
    <parent link="chassis_link"/>
    <child link="imu"/>
  </joint>
  <link name="imu"/>
  <!-- TODO(quintin): Import the ZED URDF here -->
  <joint name="chassis_link_to_zed_left_camera_frame" type="fixed">
    <origin xyz="0.3 0 0"/>
    <parent link="chassis_link"/>
    <child link="zed_left_camera_frame"/>
  </joint>
  <link name="zed_left_camera_frame">
    <collision>
      <geometry>
        <box size="0.01 0.1 0.01"/>
      </geometry>
    </collision>
  </link>
  <joint name="chassis_link_to_long_range_camera_link" type="fixed">
    <origin xyz="0.3 0 0"/>
    <parent link="chassis_link"/>
    <child link="long_range_camera_link"/>
  </joint>
  <link name="long_range_camera_link">
    <collision>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>
  </link>
  <!-- TODO(quintin): I removed this for now, feel free to add it back when needed -->
  <!--  <joint name="chassis_link_to_mast_link" type="fixed">-->
  <!--    <origin xyz="0.0505 -0.086 0.558"/>-->
  <!--    <parent link="chassis_link"/>-->
  <!--    <child link="mast_link"/>-->
  <!--  </joint>-->
  <!--  <link name="mast_link"/>-->
  <!--  <joint name="mast_link_to_zed_mini_camera" type="continuous">-->
  <!--    <origin xyz="0.05 0 0.0"/>-->
  <!--    <parent link="mast_link"/>-->
  <!--    <child link="zed_mini_camera"/>-->
  <!--    <axis xyz="0 0 1"/>-->
  <!--    <limit effort="2.68" velocity="0"/>-->
  <!--  </joint>-->
  <!--  <link name="zed_mini_camera"/>-->
</robot>
